core/vm/eof.go

```diff
@@ -0,0 +1,243 @@
+	}
+	if typesSize > 4*1024 {
+		return fmt.Errorf("number of code sections must not exceed 1024 (got %d)", typesSize)
+	}
+
+	// Parse code sizes.
+	if kind, codeSizes = parseSectionList(b, offsetCodeKind); kind != kindCode {
+		return fmt.Errorf("expected kind code")
+	}
+	if len(codeSizes) != typesSize/4 {
+		return fmt.Errorf("mismatch of code sections count and type signatures (types %d, code %d)", typesSize/3, len(codeSizes))
+	}
+
+	// Parse data size.
+	offsetDataKind := offsetCodeKind + 2 + 2*len(codeSizes) + 1
+	if len(b) < offsetDataKind+2 {
+		return fmt.Errorf("container size invalid")
+	}
+	if kind, dataSize = parseSection(b, offsetDataKind); kind != kindData {
+		return fmt.Errorf("expected kind data")
+	}
+	offsetTerminator := offsetDataKind + 3
+	if check(b, offsetTerminator, 0) {
+		return fmt.Errorf("expected terminator")
+	}
+
+	// Check for terminator.
+	expectedSize := offsetTerminator + typesSize + sum(codeSizes) + dataSize + 1
+	if len(b) != expectedSize {
+		return fmt.Errorf("invalid container size (want %d, got %d)", expectedSize, len(b))
+	}
+
+	// Parse types section.
+	idx := offsetTerminator + 1
+	var types []*FunctionMetadata
+	for i := 0; i < typesSize/4; i++ {
+		sig := &FunctionMetadata{
+			Input:          b[idx+i*4],
+			Output:         b[idx+i*4+1],
+			MaxStackHeight: uint16(parseUint16(b[idx+i*4+2:])),
+		}
+		if sig.MaxStackHeight > 1024 {
+			return fmt.Errorf("type annotation %d max stack height must not exceed 1024", i)
+		}
+		types = append(types, sig)
+	}
+	if types[0].Input != 0 || types[0].Output != 0 {
+		return fmt.Errorf("input and output of first code section must be 0")
+	}
+	c.Types = types
+
+	// Parse code sections.
+	idx += typesSize
+	code := make([][]byte, len(codeSizes))
+	for i, size := range codeSizes {
+		if size == 0 {
+			return fmt.Errorf("code section %d size must not be 0", i)
+		}
+		code[i] = b[idx : idx+size]
+		idx += size
+	}
+	c.Code = code
+
+	// Parse data section.
+	c.Data = b[idx : idx+dataSize]
+
+	return nil
+}
+
+func (c *Container) ValidateCode(jt *JumpTable) error {
+	for i, code := range c.Code {
+		if err := validateCode(code, i, c.Types, jt); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// parseSection decodes a (kind, size) pair from an EOF header.
+func parseSection(b []byte, idx int) (kind, size int) {
+	return int(b[idx]), parseUint16(b[idx+1:])
+}
+
+// parseSectionList decodes a (kind, len, []codeSize) section list from an EOF
+// header.
+func parseSectionList(b []byte, idx int) (kind int, list []int) {
+	return int(b[idx]), parseList(b, idx+1)
+}
+
+// parseList decodes a list of uint16..
+func parseList(b []byte, idx int) []int {
+	count := parseUint16(b[idx:])
+	list := make([]int, count)
+	for i := 0; i < count; i++ {
+		list[i] = parseUint16(b[idx+2*i+2:])
+	}
+	return list
+}
+
+// parseUint16 parses a 16 bit unsigned integer.
+func parseUint16(b []byte) int {
+	size := uint16(b[0]) << 8
+	size += uint16(b[1])
+	return int(size)
+}
```

Comment:
Starting with this function, I think you need to introduce errors, and then let that trickle upwards
Example
```golang
// parseUint16 parses a 16 bit unsigned integer.
func parseUint16(b []byte) (int, error) {
	if len(b) < 2{
		return io.ErrUnexpectedEOF
	}
	size := uint16(b[0]) << 8
	size += uint16(b[1])
	return int(size), nil
}
```

Reply:
Also, for reference. This method, as is, is identical (but less optimized) to `binary.BigEndian.Uint16(b)`:

```
func (bigEndian) Uint16(b []byte) uint16 {
	_ = b[1] // bounds check hint to compiler; see golang.org/issue/14808
	return uint16(b[1]) | uint16(b[0])<<8
}
```

I suspect you can replace much of this code with bounds-checks plus existing methods from `binary`